zipkin-storage/elasticsearch/src/main/java/zipkin2/elasticsearch/ElasticsearchStorage.java

```diff
@@ -265,37 +266,71 @@ void clear(String index) throws IOException {
 
   /** This is blocking so that we can determine if the cluster is healthy or not */
   @Override public CheckResult check() {
-    return ensureClusterReady(indexNameFormatter().formatType(SPAN));
+    return ensureIndexTemplatesAndClusterReady(indexNameFormatter().formatType(TYPE_SPAN));
   }
 
-  CheckResult ensureClusterReady(String index) {
+  /**
+   * This allows the health check to display problems, such as access, installing the index
+   * template. It also helps reduce traffic sent to nodes still initializing (when guarded on the
+   * check result). Finally, this reads the cluster health of the index as it can go down after the
+   * one-time initialization passes.
+   */
+  CheckResult ensureIndexTemplatesAndClusterReady(String index) {
     try {
-      HttpCall.Factory http = http();
-      AggregatedHttpRequest request = AggregatedHttpRequest.of(
-        HttpMethod.GET, "/_cluster/health/" + index);
-      return http.newCall(request, READ_STATUS, "get-cluster-health").execute();
+      ensureIndexTemplates(); // called only once, so we have to double-check health
+      AggregatedHttpRequest request = AggregatedHttpRequest.of(GET, "/_cluster/health/" + index);
+      CheckResult result = http().newCall(request, READ_STATUS, "get-cluster-health").execute();
+      if (result == null) throw new IllegalArgumentException("No content reading cluster health");
+      return result;
     } catch (Throwable e) {
       Call.propagateIfFatal(e);
-      // Unwrap the marker exception as the health check is not relevant for the throttle component,
-      // and wrapping interferes with humans intended to read this message.
-      return CheckResult.failed(e instanceof RejectedExecutionException ? e.getCause() : e);
+      // Wrapping interferes with humans intended to read this message:
+      //
+      // Unwrap the marker exception as the health check is not relevant for the throttle component.
+      // Unwrap any IOException from the first call to ensureIndexTemplates()
+      if (e instanceof RejectedExecutionException || e instanceof UncheckedIOException) {
+        e = e.getCause();
+      }
+      return CheckResult.failed(e);
     }
   }
 
-  @Memoized // since we don't want overlapping calls to apply the index templates
+  volatile IndexTemplates indexTemplates; // visible for testing
+
+  // Memoized since we don't want overlapping calls to apply the index templates
   IndexTemplates ensureIndexTemplates() {
+    if (indexTemplates == null) {
```

Comment:
this is the same thing that auto-value would generate (we do this all over in zipkin-reporter). plus on my list to un-auto-value anyway.